File: ios/CourierSharedModule.swift
import Courier_iOS

@objc(CourierSharedModule)
class CourierSharedModule: CourierReactNativeEventEmitter {
    
    private var nativeEmitters = [String]()
    private var authenticationListeners: [String: CourierAuthenticationListener] = [:]
    private var inboxListeners: [String: CourierInboxListener] = [:]

    override func stopObserving() {
        removeAuthListeners()
        removeInboxListeners()
    }
  
    @objc(attachEmitter:withResolver:withRejecter:)
    func attachEmitter(emitterId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        nativeEmitters.append(emitterId)
        resolve(emitterId)
        
    }
    
    // MARK: Client
  
    @objc(getClient:withRejecter:)
    func getClient(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
        
            guard let options = await Courier.shared.client?.options else {
                resolve(nil)
                return
            }
        
            do {
                  
                let dictionary = [
                    "jwt": options.jwt as Any,
                    "clientKey": options.clientKey as Any,
                    "userId": options.userId as Any,
                    "connectionId": options.connectionId as Any,
                    "tenantId": options.tenantId as Any,
                    "showLogs": options.showLogs as Any
                ]
                .compactMapValues { $0 }
                
                let jsonData = try JSONSerialization.data(
                    withJSONObject: dictionary,
                    options: .prettyPrinted
                )
                
                guard let jsonString = String(data: jsonData, encoding: .utf8) else {
                    resolve(nil)
                    return
                }
                
                resolve(jsonString)
                  
            } catch {
                resolve(nil)
            }
          
        }
        
    }
    
    // MARK: Authentication
    
    @objc(getUserId:withRejecter:)
    func getUserId(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            let userId = await Courier.shared.userId
            resolve(userId)
        }
    }
    
    @objc(getTenantId:withRejecter:)
    func getTenantId(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            let tenantId = await Courier.shared.tenantId
            resolve(tenantId)
        }
    }
    
    @objc(getIsUserSignedIn:withRejecter:)
    func getIsUserSignedIn(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            let isUserSignedIn = await String(Courier.shared.isUserSignedIn)
            resolve(isUserSignedIn)
        }
    }

    @objc(signIn:withClientKey:withUserId:withTenantId:withShowLogs:withResolver:withRejecter:)
    func signIn(accessToken: String, clientKey: String?, userId: String, tenantId: String?, showLogs: Bool, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        
        Task {
            
            await Courier.shared.signIn(
                userId: userId,
                tenantId: tenantId,
                accessToken: accessToken,
                clientKey: clientKey,
                showLogs: showLogs
            )
            
            resolve(nil)
            
        }
      
    }

    @objc(signOut:withRejecter:)
    func signOut(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            await Courier.shared.signOut()
            resolve(nil)
        }
    }

    @objc(addAuthenticationListener:withResolver:withRejecter:)
    func addAuthenticationListener(listenerId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        
        Task {
            
            let listener = await Courier.shared.addAuthenticationListener { [weak self] userId in
                self?.broadcast(
                    name: listenerId,
                    body: userId
                )
            }
            
            authenticationListeners[listenerId] = listener
            
            resolve(listenerId)
          
        }
        
    }
    
    @objc(removeAuthenticationListener:withResolver:withRejecter:)
    func removeAuthenticationListener(listenerId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        
        Task {
            
            let listener = authenticationListeners[listenerId]
            
            // Disable the listener
            listener?.remove()
            
            // Remove the id from the map
            authenticationListeners.removeValue(forKey: listenerId)
            
            resolve(listenerId)
          
        }
        
    }
    
    @objc(removeAllAuthenticationListeners:withRejecter:)
    func removeAllAuthenticationListeners(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            removeAuthListeners()
            resolve(nil)
        }
    }
  
    private func removeAuthListeners() {
      
        for value in authenticationListeners.values {
            value.remove()
        }
      
        authenticationListeners.removeAll()
      
    }
    
    // MARK: Push
    
    @objc(getAllTokens:withRejecter:)
    func getAllTokens(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Task {
            let tokens = await Courier.shared.tokens
            resolve(tokens)
        }
    }
    
    @objc(getToken:withResolver:withRejecter:)
    func getToken(provider: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            let token = await Courier.shared.getToken(for: provider)
            resolve(token)
        }
        
    }

    @objc(setToken:withToken:withResolver:withRejecter:)
    func setToken(provider: String, token: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.setToken(
                    for: provider,
                    token: token
                )
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
    
    }
    
    // MARK: Inbox
    
    @objc(getInboxPaginationLimit:withRejecter:)
    func getInboxPaginationLimit(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            let limit = await String(describing: Courier.shared.inboxPaginationLimit)
            resolve(limit)
        }
    }
    
    @objc(setInboxPaginationLimit:withResolver:withRejecter:)
    func setInboxPaginationLimit(limit: Double, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            await Courier.shared.setPaginationLimit(Int(limit))
            let limit = await String(describing: Courier.shared.inboxPaginationLimit)
            resolve(limit)
        }
    }
    
    @objc(openMessage:withResolver:withRejecter:)
    func openMessage(messageId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.openMessage(messageId)
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(archiveMessage:withResolver:withRejecter:)
    func archiveMessage(messageId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.archiveMessage(messageId)
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(clickMessage:withResolver:withRejecter:)
    func clickMessage(messageId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.clickMessage(messageId)
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(readMessage:withResolver:withRejecter:)
    func readMessage(messageId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.readMessage(messageId)
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(unreadMessage:withResolver:withRejecter:)
    func unreadMessage(messageId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.unreadMessage(messageId)
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(readAllInboxMessages:withRejecter:)
    func readAllInboxMessages(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task {
            do {
                try await Courier.shared.readAllInboxMessages()
                resolve(nil)
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
        
    }
    
    @objc(addInboxListener:withLoadingId:withErrorId:withUnreadCountId:withTotalCountId:withMessagesChangedId:withPageAddedId:withMessageEventId:withResolver:withRejecter:)
    func addInboxListener(
        listenerId: String,
        loadingId: String,
        errorId: String,
        unreadCountId: String,
        totalCountId: String,
        messagesChangedId: String,
        pageAddedId: String,
        messageEventId: String,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) -> Void {
      
      Task {
          let listener = await Courier.shared.addInboxListener(
              onLoading: { [weak self] isRefresh in
                  self?.broadcast(
                      name: loadingId,
                      body: isRefresh
                  )
              },
              onError: { [weak self] error in
                  self?.broadcast(
                      name: errorId,
                      body: String(describing: error)
                  )
              },
              onUnreadCountChanged: { [weak self] unreadCount in
                  self?.broadcast(
                      name: unreadCountId,
                      body: unreadCount
                  )
              },
              onTotalCountChanged: { [weak self] totalCount, feed in
                  let feedName = (feed == .archive) ? "archive" : "feed"
                  let json: [String: Any] = [
                      "totalCount": totalCount,
                      "feed": feedName
                  ]
                  self?.broadcast(
                      name: totalCountId,
                      body: json
                  )
              },
              onMessagesChanged: { [weak self] messages, canPaginate, feed in
                  // Convert each InboxMessage to JSON
                  do {
                      let feedName = (feed == .archive) ? "archive" : "feed"
                      let json: [String: Any] = [
                          "feed": feedName,
                          "canPaginate": canPaginate,
                          "messages": try messages.map { try $0.toJson() ?? "" }
                      ]
                      self?.broadcast(
                          name: messagesChangedId,
                          body: json
                      )
                  } catch {
                      Task {
                          await Courier.shared.client?.error(error.localizedDescription)
                      }
                  }
              },
              onPageAdded: { [weak self] messages, canPaginate, isFirstPage, feed in
                  // Convert each InboxMessage to JSON
                  do {
                      let feedName = (feed == .archive) ? "archive" : "feed"
                      let json: [String: Any] = [
                          "feed": feedName,
                          "canPaginate": canPaginate,
                          "isFirstPage": isFirstPage,
                          "messages": try messages.map { try $0.toJson() ?? "" }
                      ]
                      self?.broadcast(
                          name: pageAddedId,
                          body: json
                      )
                  } catch {
                      Task {
                          await Courier.shared.client?.error(error.localizedDescription)
                      }
                  }
              },
              onMessageEvent: { [weak self] message, index, feed, event in
                  do {
                      let feedName = (feed == .archive) ? "archive" : "feed"
                      let eventName = event.rawValue  // or handle explicitly as needed
                      let json: [String: Any] = [
                          "feed": feedName,
                          "event": eventName,
                          "index": index,
                          "message": try message.toJson() ?? ""
                      ]
                      self?.broadcast(
                          name: messageEventId,
                          body: json
                      )
                  } catch {
                      Task {
                          await Courier.shared.client?.error(error.localizedDescription)
                      }
                  }
              }
          )
          
          // Store the listener so you can remove it later if needed
          inboxListeners[listenerId] = listener
          
          // Resolve with the listenerId so JS knows the registration succeeded
          resolve(listenerId)
      }
    }
    
    @objc(removeInboxListener:withResolver:withRejecter:)
    func removeInboxListener(listenerId: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        
        Task {
            
            let listener = inboxListeners[listenerId]
            
            // Disable the listener
            listener?.remove()
            
            // Remove the id from the map
            inboxListeners.removeValue(forKey: listenerId)
            
            resolve(listenerId)
          
        }
        
    }
    
    @objc(removeAllInboxListeners:withRejecter:)
    func removeAllInboxListeners(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            removeInboxListeners()
            resolve(nil)
        }
    }
  
    private func removeInboxListeners() {
        for value in inboxListeners.values {
            value.remove()
        }
        inboxListeners.removeAll()
    }
    
    @objc(refreshInbox: withRejecter:)
    func refreshInbox(resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            await Courier.shared.refreshInbox()
            resolve(nil)
        }
    }
    
    @objc(fetchNextPageOfMessages:withResolver:withRejecter:)
    func fetchNextPageOfMessages(inboxMessageFeed: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Task {
            do {
                let value: InboxMessageFeed = inboxMessageFeed == "archive" ? .archive : .feed
                let messages = try await Courier.shared.fetchNextInboxPage(value)
                resolve(try messages.map { try $0.toJson() ?? "" })
            } catch {
                Rejections.sharedError(reject, error: error)
            }
        }
    }

    override func supportedEvents() -> [String]! {
        return nativeEmitters
    }

    @objc(setIsUITestsActive:)
    func setIsUITestsActive(_ isActive: Bool) {
      Courier.isUITestsActive = isActive
    }
    
}

File: android/src/main/java/com/courierreactnative/CourierSharedModule.kt
package com.courierreactnative

import com.courier.android.Courier
import com.courier.android.models.CourierAuthenticationListener
import com.courier.android.models.CourierInboxListener
import com.courier.android.models.InboxMessageSet
import com.courier.android.models.remove
import com.courier.android.modules.addAuthenticationListener
import com.courier.android.modules.addInboxListener
import com.courier.android.modules.archiveMessage
import com.courier.android.modules.clickMessage
import com.courier.android.modules.fetchNextInboxPage
import com.courier.android.modules.getToken
import com.courier.android.modules.inboxPaginationLimit
import com.courier.android.modules.isUserSignedIn
import com.courier.android.modules.openMessage
import com.courier.android.modules.readAllInboxMessages
import com.courier.android.modules.readMessage
import com.courier.android.modules.refreshInbox
import com.courier.android.modules.setToken
import com.courier.android.modules.signIn
import com.courier.android.modules.signOut
import com.courier.android.modules.tenantId
import com.courier.android.modules.tokens
import com.courier.android.modules.unreadMessage
import com.courier.android.modules.userId
import com.courier.android.ui.inbox.InboxMessageFeed
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class CourierSharedModule(
  reactContext: ReactApplicationContext
) : ReactNativeModule(
  tag = "Shared Instance Error",
  name = "CourierSharedModule",
  reactContext = reactContext
) {

  // Listeners
  private var authenticationListeners = mutableMapOf<String, CourierAuthenticationListener>()
  private var inboxListeners = mutableMapOf<String, CourierInboxListener>()

  @ReactMethod
  fun addListener(type: String?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @ReactMethod
  fun removeListeners(type: Int?) {
    // Keep: Required for RN built in Event Emitter Calls.
  }

  @ReactMethod
  fun attachEmitter(emitterId: String, promise: Promise) {
    promise.resolve(emitterId)
  }

  // Client

  @ReactMethod
  fun getClient(promise: Promise) {
    val options = Courier.shared.client?.options
    promise.resolve(options?.toJson())
  }

  // Authentication

  @ReactMethod
  fun getUserId(promise: Promise) {
    promise.resolve(Courier.shared.userId)
  }

  @ReactMethod
  fun getTenantId(promise: Promise) {
    promise.resolve(Courier.shared.tenantId)
  }

  @ReactMethod
  fun getIsUserSignedIn(promise: Promise) {
    promise.resolve(Courier.shared.isUserSignedIn.toString())
  }

  @ReactMethod
  fun signIn(
    accessToken: String,
    clientKey: String?,
    userId: String,
    tenantId: String?,
    showLogs: Boolean,
    promise: Promise
  ) {
    CoroutineScope(Dispatchers.Main).launch {
      Courier.shared.signIn(
        userId = userId,
        tenantId = tenantId,
        accessToken = accessToken,
        clientKey = clientKey,
        showLogs = showLogs
      )
      promise.resolve(null)
    }
  }

  @ReactMethod
  fun signOut(promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      Courier.shared.signOut()
      promise.resolve(null)
    }
  }

  @ReactMethod
  fun addAuthenticationListener(listenerId: String, promise: Promise) {
    // Create the listener
    val listener = Courier.shared.addAuthenticationListener { userId ->
      reactApplicationContext.sendEvent(
        eventName = listenerId,
        value = userId
      )
    }

    // Add the listener to the map
    authenticationListeners[listenerId] = listener
    promise.resolve(listenerId)
  }

  @ReactMethod
  fun removeAuthenticationListener(listenerId: String, promise: Promise) {
    val listener = authenticationListeners[listenerId]
    listener?.remove()
    authenticationListeners.remove(listenerId)
    promise.resolve(listenerId)
  }

  @ReactMethod
  fun removeAllAuthenticationListeners(promise: Promise) {
    authenticationListeners.values.forEach { it.remove() }
    authenticationListeners.clear()
    promise.resolve(null)
  }

  // Push

  @ReactMethod
  fun getAllTokens(promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      val tokens = Courier.shared.tokens
      val resultMap = Arguments.createMap()
      tokens.forEach { (key, value) ->
        resultMap.putString(key, value)
      }
      promise.resolve(resultMap)
    }
  }

  @ReactMethod
  fun getToken(provider: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      val token = Courier.shared.getToken(provider)
      promise.resolve(token)
    }
  }

  @ReactMethod
  fun setToken(provider: String, token: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.setToken(provider, token)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  // Inbox

  @ReactMethod
  fun getInboxPaginationLimit(promise: Promise) {
    promise.resolve(Courier.shared.inboxPaginationLimit.toString())
  }

  @ReactMethod
  fun setInboxPaginationLimit(limit: Double, promise: Promise) {
    Courier.shared.inboxPaginationLimit = limit.toInt()
    promise.resolve(Courier.shared.inboxPaginationLimit.toString())
  }

  @ReactMethod
  fun openMessage(messageId: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.openMessage(messageId)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun archiveMessage(messageId: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.archiveMessage(messageId)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun clickMessage(messageId: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.clickMessage(messageId)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun readMessage(messageId: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.readMessage(messageId)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun unreadMessage(messageId: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.unreadMessage(messageId)
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun readAllInboxMessages(promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        Courier.shared.readAllInboxMessages()
        promise.resolve(null)
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }

  @ReactMethod
  fun addInboxListener(
    listenerId: String,
    loadingId: String,
    errorId: String,
    unreadCountId: String,
    totalCountId: String,
    messagesChangedId: String,
    pageAddedId: String,
    messageEventId: String,
    promise: Promise
  ) {
    CoroutineScope(Dispatchers.Main).launch {
      val listener = Courier.shared.addInboxListener(
        onLoading = { isRefresh ->
          reactApplicationContext.sendEvent(
            eventName = loadingId,
            value = isRefresh
          )
        },
        onError = { e ->
          reactApplicationContext.sendEvent(
            eventName = errorId,
            value = e.message ?: "Courier Inbox Error"
          )
        },
        onUnreadCountChanged = { unreadCount ->
          reactApplicationContext.sendEvent(
            eventName = unreadCountId,
            value = unreadCount
          )
        },
        onTotalCountChanged = { totalCount, feed ->
          val json = Arguments.createMap().apply {
            putString("feed", if (feed == InboxMessageFeed.FEED) "feed" else "archive")
            putInt("totalCount", totalCount)
          }
          reactApplicationContext.sendEvent(
            eventName = totalCountId,
            value = json
          )
        },
        onMessagesChanged = { messages, canPaginate, feed ->
          val json = Arguments.createMap().apply {
            putString("feed", if (feed == InboxMessageFeed.FEED) "feed" else "archive")
            putArray("messages", messages.map { it.toJson() }.toWritableArray())
            putBoolean("canPaginate", canPaginate)
          }
          reactApplicationContext.sendEvent(
            eventName = messagesChangedId,
            value = json
          )
        },
        onPageAdded = { messages, canPaginate, isFirstPage, feed ->
          val json = Arguments.createMap().apply {
            putString("feed", if (feed == InboxMessageFeed.FEED) "feed" else "archive")
            putArray("messages", messages.map { it.toJson() }.toWritableArray())
            putBoolean("canPaginate", canPaginate)
            putBoolean("isFirstPage", isFirstPage)
          }
          reactApplicationContext.sendEvent(
            eventName = pageAddedId,
            value = json
          )
        },
        onMessageEvent = { message, index, feed, event ->
          val json = Arguments.createMap().apply {
            putString("feed", if (feed == InboxMessageFeed.FEED) "feed" else "archive")
            putInt("index", index)
            putString("event", event.value)
            putString("message", message.toJson())
          }
          reactApplicationContext.sendEvent(
            eventName = messageEventId,
            value = json
          )
        }
      )

      inboxListeners[listenerId] = listener
      promise.resolve(listenerId)
    }
  }

  private fun InboxMessageSet.toJson(): WritableMap? {
    val json = Arguments.createMap()
    json.putArray("messages", messages.toList().map { it.toJson() }.toWritableArray())
    json.putInt("totalMessageCount", totalCount)
    json.putBoolean("canPaginate", canPaginate)
    return json
  }

  @ReactMethod
  fun removeInboxListener(listenerId: String, promise: Promise) {
    val listener = inboxListeners[listenerId]
    listener?.remove()
    inboxListeners.remove(listenerId)
    promise.resolve(listenerId)
  }

  @ReactMethod
  fun removeAllInboxListeners(promise: Promise) {
    inboxListeners.values.forEach { it.remove() }
    inboxListeners.clear()
    promise.resolve(null)
  }

  @ReactMethod
  fun refreshInbox(promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      Courier.shared.refreshInbox()
      promise.resolve(null)
    }
  }

  @ReactMethod
  fun fetchNextPageOfMessages(inboxMessageFeed: String, promise: Promise) {
    CoroutineScope(Dispatchers.Main).launch {
      try {
        val messageSet = Courier.shared.fetchNextInboxPage(
          if (inboxMessageFeed == "archived") InboxMessageFeed.ARCHIVE else InboxMessageFeed.FEED
        )
        promise.resolve(messageSet?.toJson())
      } catch (e: Exception) {
        promise.apiError(e)
      }
    }
  }
}
